src/fido/types.h

```diff
@@ -16,12 +16,18 @@ extern "C" {
 
 struct fido_dev;
 
-typedef void *fido_dev_io_open_t(const char *);
-typedef void  fido_dev_io_close_t(void *);
-typedef int   fido_dev_io_read_t(void *, unsigned char *, size_t, int);
-typedef int   fido_dev_io_write_t(void *, const unsigned char *, size_t);
-typedef int   fido_dev_rx_t(struct fido_dev *, uint8_t, unsigned char *, size_t, int);
-typedef int   fido_dev_tx_t(struct fido_dev *, uint8_t, const unsigned char *, size_t);
+typedef struct fido_dev_io_info {
+	void *io_handle;
+	uint16_t report_in_len;
+	uint16_t report_out_len;
+} fido_dev_io_info_t;
+
+typedef struct fido_dev_io_info *fido_dev_io_open_t(const char *);
+typedef void   fido_dev_io_close_t(struct fido_dev_io_info *);
+typedef int    fido_dev_io_read_t(struct fido_dev_io_info *, unsigned char *, size_t, int);
+typedef int    fido_dev_io_write_t(struct fido_dev_io_info *, const unsigned char *, size_t);
+typedef int    fido_dev_rx_t(struct fido_dev *, uint8_t, unsigned char *, size_t, int);
+typedef int    fido_dev_tx_t(struct fido_dev *, uint8_t, const unsigned char *, size_t);
```

Comment:
These definitions are part of the public API. Would it be possible to keep them as they are, and cast `void *` to `fido_dev_io_info_t` internally, similar to what hid_openbsd.c and hid_osx.c do?

Reply:
Thanks for catching that, I wasn't aware that these functions were part of the public API until now. Yes, that should be possible, I will force-push a new version once I have it ready.

Reply:
I have amended the first commit to preserve the public API. This is done by including a flag in `fido_dev_t` that allows switching between the native, report size aware io function signature expecting a `fido_dev_io_info_t` and the legacy io functions expecting just a void pointer. If desired, another commit could add a version of `fido_dev_set_io_functions` that is report size aware.